File: pasterino/src/main/java/com/pyamsoft/pasterino/dagger/PasterinoModule.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pasterino.dagger;

import android.content.Context;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import com.pyamsoft.pasterino.PasterinoPreferences;
import com.pyamsoft.pasterino.dagger.main.MainSettingsModule;
import com.pyamsoft.pasterino.dagger.service.PasteServiceModule;

public class PasterinoModule {

  @NonNull private final Provider provider;

  public PasterinoModule(@NonNull Context context) {
    provider = new Provider(context);
  }

  @CheckResult @NonNull public final PasteServiceModule providePasteServiceModule() {
    return new PasteServiceModule(provider);
  }

  @CheckResult @NonNull public final MainSettingsModule provideMainSettingsModule() {
    return new MainSettingsModule(provider);
  }

  public static class Provider {

    @NonNull private final PasterinoPreferences preferences;

    Provider(final @NonNull Context context) {
      preferences = new PasterinoPreferencesImpl(context.getApplicationContext());
    }

    @CheckResult @NonNull public final PasterinoPreferences providePreferences() {
      return preferences;
    }
  }
}

File: pasterino-base/src/main/java/com/pyamsoft/pasterino/base/PasterinoModule.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.pasterino.base

import android.support.annotation.CheckResult
import com.pyamsoft.pasterino.base.preference.ClearPreferences
import com.pyamsoft.pasterino.base.preference.PastePreferences
import com.pyamsoft.pasterino.model.ServiceEvent
import com.pyamsoft.pydroid.PYDroidModule
import com.pyamsoft.pydroid.bus.EventBus
import com.pyamsoft.pydroid.loader.ImageLoader
import com.pyamsoft.pydroid.loader.LoaderModule
import io.reactivex.Scheduler

class PasterinoModule(private val pyDroidModule: PYDroidModule,
        private val loaderModule: LoaderModule) {

    private val preferences: PasterinoPreferencesImpl = PasterinoPreferencesImpl(
            pyDroidModule.provideContext())
    private val pasteBus = PasteBus()

    @CheckResult
    fun providePasteBus(): EventBus<ServiceEvent> = pasteBus

    @CheckResult
    fun providePreferences(): PastePreferences = preferences

    @CheckResult
    fun provideClearPreferences(): ClearPreferences = preferences

    @CheckResult
    fun provideMainScheduler(): Scheduler = pyDroidModule.provideMainThreadScheduler()

    @CheckResult
    fun provideIoScheduler(): Scheduler = pyDroidModule.provideIoScheduler()

    @CheckResult
    fun provideComputationScheduler(): Scheduler = pyDroidModule.provideComputationScheduler()

    @CheckResult
    fun provideImageLoader(): ImageLoader = loaderModule.provideImageLoader()
}
